File: haskell/pro5.hs
helper :: Int -> [Int] -> Int
helper n lst =
  foldl (\n d -> if 0 == n `mod` d then n `div` d else n) n lst

pro5 :: Int -> Int
pro5 n =
  let rec m lst =
        if m >= n
        then product lst
        else rec (m + 1) (helper m lst : lst)
  in rec 2 [1]

File: python/pro5.py
# -*- coding: utf-8 -*-

def tmp(n, d):
    if 0 == n % d:
        return n / d
    else:
        return n

def helper(n, lst):
    return reduce(tmp, lst, n)

def pro5(n):
    m = 2
    lst = [1]
    while m < n:
        m = m + 1
        lst.append(helper(m, lst))
    return reduce(lambda a, b: a * b, lst, 1)

print pro5(20)
